ajna-finance/ecosystem-coordination
# Delegation Reward implemention and bug fixes

1. Fix check for unbounded negative budget allocation in funding period of standard proposal.
https://www.notion.so/worlds-best-crypto-lending-platform/Grant-Coordination-Fund-Implement-delegation-rewards-874b2e175605448685926dd6a1ff3162
2. Fix check for multiple votes by a single voter in screening round.
3. Implement delegatee rewards. Reward is calculated on basis of their vote share in funding stage. After challenge period has ended delegatee can claim his reward by calling `claimDelegateReward` method.
4. Update `hasVoted` method that is an override of Governor contract.
5. No proposal creation after end of screening period for a particular distribution.
https://www.notion.so/worlds-best-crypto-lending-platform/Validate-distributionId-associated-with-proposeStandard-2c4e3268bfc048df81084aa536fc6a95
6. Execute standard should be callable even after a new distribution has been created. 
https://www.notion.so/worlds-best-crypto-lending-platform/Fix-GrantCoordinationFund-executeStandard-bug-bbe2d7a666cd4de294fadbeca126a8d4
7. Test coverage improvements

```
Before 
                            Line coverage       Function coverage       Branch coverage
GrantFund.sol               87.5 %                50.0%                    85.7 %
ExtraordinaryFunding.sol    97.3 %                100.0%                    81.2 %
Funding.sol                    92.3 %                50.0%                    83.3 %
StandardFunding.sol            96.3 %                90.5%                    73.3 %


After
                            Line coverage       Function coverage       Branch coverage
GrantFund.sol               95.5 %                86.7 %                    93.5 %
ExtraordinaryFunding.sol    100.0 %                100.0 %                    100.0 %
Funding.sol                    100.0 %                100.0 %                    100.0 %
StandardFunding.sol            100.0 %                100.0 %                    90.5 %
```
8. Add fuzz test for top ten proposals. Fixed a bug in top ten array updation.
9. Add fuzz test for delegation rewards. Fixed a potential attack vector by updating proposals array to proposal Id array in check slate.
10. Add fuzz test for EFM. Fix code to use non-treasury token votes intead of treasury token votes.
